src/lightning/pytorch/trainer/connectors/data_connector.py

```diff
@@ -442,14 +442,11 @@ def _worker_check(dataloader: object, using_spawn: bool, name: str) -> None:
                 "strategy=ddp_spawn and num_workers=0 may result in data loading bottlenecks."
                 " Consider setting num_workers>0 and persistent_workers=True"
             )
-
-    elif dataloader.num_workers <= 2 < num_cpus and not using_spawn:
+    elif dataloader.num_workers <= 2 < upper_bound or dataloader.num_workers < 2 <= upper_bound:
```

Comment:
the first part is pretty much certain to be True on the majority of setups I think, since the default in many frameworks is usually 2 workers and most modern machines have lots of cpu-cores for desktops, and any serious gpus nodes will have lots of cpu-cores.

Would you consider that `num_workers == 2` is actually a reasonable setting?

Reply:
If I read you correctly, you suggest simplifying the condition by dropping `2 < upper_bound` because it's just true on most DL systems. I think I agree. So we would be left with only this check:

dataloader.num_workers <= 2

Now the question is, should this become a different upper limit. The history here is that when this was added back when PL was developed, the majority of users were doing computer vision and num workers > 2 was almost always mandatory for applying augmentations.

Today if we do small to medium size LLM training on small pre-tokenized datasets that fit in a machine, we're not going to require that many workers. So num workers=2 is not that unreasonable. We could consider dropping that to 

dataloader.num_workers < 2

to emit the warning.

Reply:
We are in agreement, Adrian

- language models, especially where data is preprocessed, ala Megatron-LM/nemo - 2 workers is the norm from my experience
- vision models is a different story, if the amount of transforms is huge it can be very slow and yes warrant more workers

Ideally the general solution would be to provide users with a breakdown of timing for `[DL][fwd+bwd][logging]` so that they can see if their DL is a bottleneck. I include logging as well since it can also be an issue if one does some blocking logging.

Now you're switching to measuring the real impact of num_workers to the efficiency of the compute.

So for example you could warn a user if you see that dl_time > 10% of compute time or something like that. I don't have a good general recommendation on threshold % since LLM/VLM workload could be very different to a 1 gpu workload. But the principle is the same - compute is the most expensive resource, especially if it's rented per hour, so now the user wants to detect and remove the bottlenecks to pay less and of course have an earlier finish line.

----------------

Now, with the raise of the cloud storage there is a new problem emerging and that is of the data not even being present on the compute node at the time of compute. So not only DL might need to do transforms, it also needs to fetch remove data from the cloud. Here 2 workers are almost never enough. During IDEFICS-80B training we had this issue, but we couldn't raise the number of workers not because we didn't have the spare cores, but because WebDataset which we were using for Just-in-time prefetch lead to huge processes so that the 2x8 workers were consuming a lion part of 1+TB of memory and so with even 3 workers we would get cgroups killing the training.

I hope that WebDataset and other streaming DL solutions will get better over time, but this was a very painful experience 6 months ago, since we did want more workers, but couldn't have them.